### ISSUE TITLE
Internal error. Debug Failure. False expression.

### ISSUE THREAD

**Issue by pawelrubin** [microsoft/pyright#1114]
**Describe the bug**
**Pyright** crashes in the code below.

(This is exported from a huge codebase. I've done what I could to reproduce it in the more general case)

**To Reproduce**
Consider this piece of code
```python
from __future__ import annotations

from typing import Any, Callable, Type

from blinker.base import Signal as Event
from mongoengine import Document, signals


ApplySignal = Callable[[Type[Document]], Type[Document]]
Signal = Callable[[Type[Document], Document], None]
SignalHandler = Callable[[Signal], ApplySignal]


def handler(event: Event) -> SignalHandler:
    """Signal decorator to allow use of callback functions as class decorators."""

    def decorator(func: Signal) -> ApplySignal:
        def apply(cls: Type[Document]) -> Type[Document]:
            event.connect(func, sender=cls)
            return cls

        return apply

    return decorator


@handler(signals.pre_save)
def update_foo(_: Any, document: Foo) -> None:
    return None


@update_foo  # <- there is no crash if this line is not present
class Foo(Document):
    pass

```

**Additional context**
Error details
```
An internal error occurred while type checking file "example.py": Error: Debug Failure. False expression.
	at Dt (/home/prubin/.vscode/extensions/ms-pyright.pyright-1.1.80/dist/server.js:2:579437)
    at /home/prubin/.vscode/extensions/ms-pyright.pyright-1.1.80/dist/server.js:2:613508
    at un (/home/prubin/.vscode/extensions/ms-pyright.pyright-1.1.80/dist/server.js:2:613585)
    at pn (/home/prubin/.vscode/extensions/ms-pyright.pyright-1.1.80/dist/server.js:2:613179)
    at Ke (/home/prubin/.vscode/extensions/ms-pyright.pyright-1.1.80/dist/server.js:2:534967)
    at He (/home/prubin/.vscode/extensions/ms-pyright.pyright-1.1.80/dist/server.js:2:528670)
    at /home/prubin/.vscode/extensions/ms-pyright.pyright-1.1.80/dist/server.js:2:523411
    at Object.l [as doForSubtypes] (/home/prubin/.vscode/extensions/ms-pyright.pyright-1.1.80/dist/server.js:2:656206)
    at je (/home/prubin/.vscode/extensions/ms-pyright.pyright-1.1.80/dist/server.js:2:516200)
    at re (/home/prubin/.vscode/extensions/ms-pyright.pyright-1.1.80/dist/server.js:2:487175)Pyright
```

**Comment by erictraut** [comment]
Thanks for reporting — and for providing the detailed repro.

I've fixed the crash, but there are still (legitimate) errors reported with this code.

The reason this was crashing is that the code is creating a circular dependency between a class (`Foo`) and the class decorator that decorates this class (`update_foo`). There's really no way to analyze such a circular dependency because the type of the symbol `Foo` cannot be determined until the decorator is analyzed, but the decorator's signature relies on the symbol `Foo`. It's similar to a type alias that refers to itself.

**Comment by erictraut** [comment]
This is addressed in 1.1.82, which I just published.
